### ISSUE TITLE
Bypassing Image Verification to Login?!

### ISSUE THREAD

**Issue by tradingwithme** [tedchou12/webull#367]
I am facing a conundrum: Is there anyone out there who can shed some light on this vexing issue? I am trying to resolve the verify picture verification, but alas, my attempts with the email address have been fruitless. Instead, I have been directed to use my phone number, yet every time I try, I am met with a frustrating error message on the Jupyter terminal. Can anyone help me navigate and find a solution to this problem?

![image](https://user-images.githubusercontent.com/121399099/209477785-1e0689c8-a773-4c3a-98bd-6a3c8ecf5148.png)

![image](https://user-images.githubusercontent.com/121399099/209477788-616a8350-bb76-475a-919f-681687c934ec.png)

Alternatively, I tried the "Workaround Login" method. However, it seems that the html document for Webull has changed since the last screenshot was taken by @tedchou12 . Therefore, the keywords are no longer searchable.

**Comment by tedchou12** [comment]
Thanks for the feedback!! Using `accesstoken` should find it.

**Comment by tradingwithme** [comment]
> Thanks for the feedback!! Using `accesstoken` should find it.

Thank you for your reply. I still have encountered another roadblock after attempting your suggestion.

I was able to obtain the 'access_token' information from the HAR file I downloaded from the "https://app.webull.com/paper" url. 

Here is the data that I called to become read:
![image](https://user-images.githubusercontent.com/121399099/209764804-a17c6cd8-2842-4021-8cae-404bf137cf76.png)

Afterward, I ran the following code statement(s), but to no avail:

![image](https://user-images.githubusercontent.com/121399099/209764958-86a661d5-1faa-4526-b903-39578aa775be.png)

**What would you suggest I do to resolve this issue?**

**Comment by tedchou12** [comment]
sure, let me check it latter today

**Comment by tradingwithme** [comment]
> sure, let me check it latter today

Thanks so much. Please, keep me posted.

**Comment by tradingwithme** [comment]
> Here is exactly what I did: open /usr/local/lib/python3.9/dist-packages/webull/endpoints.py, change the self.base_userfintech_url endpoint to: self.base_userfintech_url = 'https://nauser.webullfintech.com/api'
> 
> Open an incognito browser window, inspect. Login to webull. Look for the POST to the same url. Copy the deviceid and device name.
> 
> removed did.bin In my login code: wb = webull() wb._did = "DeviceIDHERE" wb.accountType = '2' data = wb.login('username@here', 'PASShere', device_name="DeviceNAMEHERE", save_token=True) print("\n\n\n\n"+ str(data) + "\n\n\n\n\n")
> 
> Then in my other code, I use this so I don't have to login on each run: wb = webull() wb._access_token = 'AccessToken_from_login_above'
> 
> This all may not be correct, I'm just trying to help since I was having the same problems and this resolved it.

@tedchou12 & @shaggy63 

Thank you for your reply. I still have encountered another roadblock after attempting your suggestion.

I have tried your suggestion as follows:
`
pwb = paper_webull() 
endpoints.base_userfintech_url = 'https://nauser.webullfintech.com/api'
pwb._did = 'did_value'
pwb.accountType = '2'
pwb._access_token = 'access_token_value'
data = pwb.login('phone_number', 'password', 'DeviceNAMEHERE', save_token=True)
print("\n\n\n\n"+ str(data) + "\n\n\n\n\n")
`
However, it printed the following message:
![image](https://user-images.githubusercontent.com/121399099/209857503-592653e6-922b-4484-bdfc-fc9c79d17e81.png)

**What would you suggest I do to resolve this issue?**

**Comment by shaggy63** [comment]
Did you edit the actual endpoints.py file that was installed with this package?

**Comment by tradingwithme** [comment]
> Did you edit the actual endpoints.py file that was installed with this package?
@tedchou12 @shaggy63 

Thank you for your reply. How so?

**Comment by tedchou12** [comment]
@tradingwithme 

<img width="1276" alt="截圖 2022-12-30 上午9 20 01" src="https://user-images.githubusercontent.com/10494709/210023538-da89694f-01a9-4e8f-989a-c21b9d90114a.png">

If you save the json in the right of the image as a json file, it should automatically load in the vars without you doing anything else:

```
from webull_package.webull import webull, paper_webull
import json


wb = paper_webull()

fh = open('webull_credentials.json', 'r')
credential_data = json.load(fh)
fh.close()
#
# print(credential_data)
wb._refresh_token = credential_data['refreshToken']
wb._access_token = credential_data['accessToken']
wb._token_expire = credential_data['tokenExpireTime']
wb._uuid = credential_data['uuid']

n_data = wb.refresh_login()
# print(n_data)

credential_data['refreshToken'] = n_data['refreshToken']
credential_data['accessToken'] = n_data['accessToken']
credential_data['tokenExpireTime'] = n_data['tokenExpireTime']

file = open('webull_credentials.json', 'w')
json.dump(credential_data, file)
file.close()


print(wb.get_account())
```

**Comment by osaidz** [comment]
Hi, I'm am facing the exact same issue on a new account. 

Access token is only searchable on https://www.webull.com/center but refresh_token and token_expire are missing so I am unable to login using the api. 

Could this be because I have not opened a full account (Non-US citizen)? I can however see quotes and level 2 data.

**Comment by tedchou12** [comment]
@osaidz 
Are you able to find it here?
https://app.webull.com/

**Comment by osaidz** [comment]
> @osaidz Are you able to find it here? https://app.webull.com/

No

![at](https://user-images.githubusercontent.com/19634325/210025969-cd27a710-b33e-4100-a721-ac4ea6b533c3.png)

**Comment by tradingwithme** [comment]
> > @osaidz Are you able to find it here? https://app.webull.com/
> 
> No
> 
> ![at](https://user-images.githubusercontent.com/19634325/210025969-cd27a710-b33e-4100-a721-ac4ea6b533c3.png)

Try typing the keyword 'access_token" on the search engine. It should return a result.

**Comment by osaidz** [comment]
@tedchou12 Thanks, 

It does return a lot of results all pointing to the same variable but I'm to find this string mentioned in 

https://github.com/tedchou12/webull/wiki/Workaround-for-Login

`{"extInfo":{"userPwdFlag":"1","equipmentCheck":"success"},"accessToken":"dc_us1.188awefa4-00000000000000000000000000000000","uuid":"165dbaa7333aeawf312151ar1f1dc804b9b","refreshToken":"188a0awef2444-00000000000000000000000000000000","tokenExpireTime":"2021-02-14T13:35:35.780+0000","firstTimeOfThird":false,"registerAddress":169,"settings":{"id":94551253056757760,"userId":300431177,"regionId":169,"language":"en","focusMarketId":"2,3,4,5,6,14","theme":2,"increDecreColor":2,"fontSize":"M","portfolioDisplayMode":2,"portfolioNameNewline":1,"portfolioHoldingsDisplay":1,"portfolioIndexDisplay":1,"portfolioBulletin":1,"kdata":1,"refreshFrequency":1,"shock":0,"tickerPriceRemind":1,"orderDealRemind":1,"hotNews":1,"chartOption":2,"operateTime":"1969-12-31T00:00:00.000+0000","languageUpdateTime":"1970-01-01T00:00:00.000+0000","createTime":"2018-09-15T05:18:25.000+0000","updateTime":"2018-09-15T05:18:25.000+0000"}}`

**Comment by tradingwithme** [comment]
> 
Just found them for you:

"tokenExpireTime":"2021-02-14T13:35:35.780+0000"
"refreshToken":"188a0awef2444-00000000000000000000000000000000"
"accessToken":"dc_us1.188awefa4-00000000000000000000000000000000","uuid":"165dbaa7333aeawf312151ar1f1dc804b9b"

FYI: Do not share this kind of information. It's hackable.

**Comment by osaidz** [comment]
> > 
> 
> Just found them for you:
> 
> "tokenExpireTime":"2021-02-14T13:35:35.780+0000" "refreshToken":"188a0awef2444-00000000000000000000000000000000" "accessToken":"dc_us1.188awefa4-00000000000000000000000000000000","uuid":"165dbaa7333aeawf312151ar1f1dc804b9b"
> 
> FYI: Do not share this kind of information. It's hackable.

Thanks! but I copied this string from the original page here  https://github.com/tedchou12/webull/wiki/Workaround-for-Login. 

This is what I see. 
![image](https://user-images.githubusercontent.com/19634325/210027524-4f197ce7-192e-46ac-9846-937b8b3c1be4.png)

if I take part of that string and search it then I do not get any results. 

![image](https://user-images.githubusercontent.com/19634325/210027575-4a11c1f4-2efa-4936-9d5c-a1a1070dc4dc.png)

**Comment by tradingwithme** [comment]
> @tradingwithme
> 
> <img alt="截圖 2022-12-30 上午9 20 01" width="1276" src="https://user-images.githubusercontent.com/10494709/210023538-da89694f-01a9-4e8f-989a-c21b9d90114a.png">
> 
> If you save the json in the right of the image as a json file, it should automatically load in the vars without you doing anything else:
> 
> ```
> from webull_package.webull import webull, paper_webull
> import json
> 
> 
> wb = paper_webull()
> 
> fh = open('webull_credentials.json', 'r')
> credential_data = json.load(fh)
> fh.close()
> #
> # print(credential_data)
> wb._refresh_token = credential_data['refreshToken']
> wb._access_token = credential_data['accessToken']
> wb._token_expire = credential_data['tokenExpireTime']
> wb._uuid = credential_data['uuid']
> 
> n_data = wb.refresh_login()
> # print(n_data)
> 
> credential_data['refreshToken'] = n_data['refreshToken']
> credential_data['accessToken'] = n_data['accessToken']
> credential_data['tokenExpireTime'] = n_data['tokenExpireTime']
> 
> file = open('webull_credentials.json', 'w')
> json.dump(credential_data, file)
> file.close()
> 
> 
> print(wb.get_account())
> ```
@tedchou12 
The keywords are different compared to mine because my account is based off the U.S. I found a search result that returned as "access_token" instead of "accessToken." Are there any other terms that I could find compatible to my JSON data?

**Comment by tedchou12** [comment]
Sorry, I am not sure about that. May be it doesn't work for newer accounts that have not been fully activated?
>Could this be because I have not opened a full account (Non-US citizen)? I can however see quotes and level 2 data.
How did you sign up?

**Comment by tradingwithme** [comment]
> > > 
> > 
> > 
> > Just found them for you:
> > "tokenExpireTime":"2021-02-14T13:35:35.780+0000" "refreshToken":"188a0awef2444-00000000000000000000000000000000" "accessToken":"dc_us1.188awefa4-00000000000000000000000000000000","uuid":"165dbaa7333aeawf312151ar1f1dc804b9b"
> > FYI: Do not share this kind of information. It's hackable.
> 
> Thanks! but I copied this string from the original page here https://github.com/tedchou12/webull/wiki/Workaround-for-Login.
> 
> This is what I see. ![image](https://user-images.githubusercontent.com/19634325/210027524-4f197ce7-192e-46ac-9846-937b8b3c1be4.png)
> 
> if I take part of that string and search it then I do not get any results.
> 
> ![image](https://user-images.githubusercontent.com/19634325/210027575-4a11c1f4-2efa-4936-9d5c-a1a1070dc4dc.png)

Yup. The format for that entry matches mine.

**Comment by osaidz** [comment]
> Sorry, I am not sure about that. May be it doesn't work for newer accounts that have not been fully activated?
> 
> > Could this be because I have not opened a full account (Non-US citizen)? I can however see quotes and level 2 data.
> 
> How did you sign up?

I signed up using my phone number then linked my email but I was not able to proceed further because it says I need to be a US resident. I'm based in the UK. 

I can only view the markets and level 2 quotes.

**Comment by tedchou12** [comment]
Unfortunately, I cant signup with another phone number (Ive only got one)
May be 
https://github.com/tedchou12/webull/issues/367#issuecomment-1367686829
would work?

**Comment by osaidz** [comment]
@tedchou12 There might be a bug becasue wb.is_logged_in() is returning true, if I run the code below. 

```
from webull import webull, StreamConn
wb = webull()
wb._access_token = 'some_token'
wb.is_logged_in()
```

**Comment by tradingwithme** [comment]
> @tedchou12 There might be a bug becasue wb.is_logged_in() is returning true, if I run the code below.
> 
> ```
> from webull import webull, StreamConn
> wb = webull()
> wb._access_token = 'some_token'
> wb.is_logged_in()
> ```

You might have to forgo implementing the 'wb.login' code statement. Instead, try 'wb.get_account_ID()' to see if it would return anything. If so, then you confirmed that you are indeed logged-in.

**Comment by SweetNaterTater** [comment]
Hey all,

I recently started playing with this package for trying to create an automated trading bot, for eventually having a bot that auto trades options by the barUpDn strategy on a paper account. I was also having issues with being able to login to my account, but you need to make sure you have selected "inspect". Then press network, so that you can begin recording any network activity that occurs when performing any actions on the web page. Then simply login to your account with your credentials, and entering your token(6 digit code for mfa). Then you should be able to find a fetched item that contains the token information.

**Comment by osaidz** [comment]
@tradingwithme 
Yes! I can confirm that I am logged in.

**Comment by tradingwithme** [comment]
> @tradingwithme Yes, I think I am logged in?
> 
> if I do wb.get_account()
> 
> I get
> 
> `{'netLiquidation': '0.00', 'unrealizedProfitLoss': '0.00', 'unrealizedProfitLossRate': '0.0000', 'accounts': [], 'positions': [], 'openOrders': [], 'openOrderSize': 0, 'positions2': [], 'openOrders2': [], 'openIpoOrders': []}`
> 
> if I do wb.refresh_login()
> 
> I get
> 
> ```
> {'msg': 'AccessToken is invalid',
>  'traceId': 'ac4dcee7c7964191ae0566d2a0376f4c',
>  'code': 'auth.token.req'}
> ```
I only know what I tried. You could try assigning the 'did' value to 'wb._did' to see if that would work. 

I did not get the output from the 'wb.get_account()  ' code statement implementation. However, I obtained a value instead from implementing the following: print(wb.get_account()). You might not need to implement the wb.refresh_login() once you run 'wb.get_account()' because what it does, I presume, is that it would cause your login status to refresh itself.

**Comment by tradingwithme** [comment]
> Hey all,
> 
> I recently started playing with this package for trying to create an automated trading bot, for eventually having a bot that auto trades options by the barUpDn strategy on a paper account. I was also having issues with being able to login to my account, but you need to make sure you have selected "inspect". Then press network, so that you can begin recording any network activity that occurs when performing any actions on the web page. Then simply login to your account with your credentials, and entering your token(6 digit code for mfa). Then you should be able to find a fetched item that contains the token information.

Does this refer to the access_token value available on the Inspect window?

**Comment by osaidz** [comment]
> > @tradingwithme Yes, I think I am logged in?
> > if I do wb.get_account()
> > I get
> > `{'netLiquidation': '0.00', 'unrealizedProfitLoss': '0.00', 'unrealizedProfitLossRate': '0.0000', 'accounts': [], 'positions': [], 'openOrders': [], 'openOrderSize': 0, 'positions2': [], 'openOrders2': [], 'openIpoOrders': []}`
> > if I do wb.refresh_login()
> > I get
> > ```
> > {'msg': 'AccessToken is invalid',
> >  'traceId': 'ac4dcee7c7964191ae0566d2a0376f4c',
> >  'code': 'auth.token.req'}
> > ```
> 
> I only know what I tried. You could try assigning the 'did' value to 'wb._did' to see if that would work.
> 
> I did not get the output after running the 'wb.get_account() ' code statement, as shown on your post. However, I obtained a value instead from implementing the following: print(wb.get_account()). You might not need to implement the wb.refresh_login() once you run 'wb.get_account()' because what it does, I presume, is that it would cause your login status to refresh itself.
> 
> Alternatively, you could try replicating what I did here to see if you get a similar result: [#367 (comment)](https://github.com/tedchou12/webull/issues/367#issuecomment-1366922522).
> 
> Do you have any existing orders or whatnot?

Thanks for helpful response. I'm using a jupyter notebook on vscode and have developed a bad habit of leaving out print statements while testing code because it does that for you automatically.

So if I do a print(wb.get_account_id()) I get None as a result. I'll try https://github.com/tedchou12/webull/issues/367#issuecomment-1366922522.

**Comment by tradingwithme** [comment]
I am using the Jupyter software from Julia.

**Comment by osaidz** [comment]
@tradingwithme I can confirm that api_login() works as I can see my alerts list functions such as get_account() or get_detail() do not return errors. 

`{'msg': 'AccessToken is expire', 'traceId': 'e0aaf937bbef4e8bab8b14dfab451d34', 'code': 'auth.token.expire'}
`

`{'msg': 'Failed to convert value of type \'java.lang.String\' to required type \'java.lang.Long\'; nested exception is java.lang.NumberFormatException: For input string: "None"', 'traceId': 'a7aa94bed56e4de697dba755c17adef4', 'code': '400', 'success': False}`

**Comment by SweetNaterTater** [comment]
> > Hey all,
> > I recently started playing with this package for trying to create an automated trading bot, for eventually having a bot that auto trades options by the barUpDn strategy on a paper account. I was also having issues with being able to login to my account, but you need to make sure you have selected "inspect". Then press network, so that you can begin recording any network activity that occurs when performing any actions on the web page. Then simply login to your account with your credentials, and entering your token(6 digit code for mfa). Then you should be able to find a fetched item that contains the token information.
> 
> Does this refer to the access_token value available on the Inspect window?

Good morning,

Yes it does, because I also was not getting the accesstoken for the webull credentials.json file. I was having issues because I was not successfully letting the inspect network capture my login attempt into webull itself. Once you have done this, you should be able to obtain the access token to bypass the mfa authentication and access either your main webull account or paper account. Hopefully this was helpful in some shape or form.

**Comment by tradingwithme** [comment]
> @tradingwithme I can confirm that api_login() works as I can see my alerts list but functions such as get_account() or get_detail() return errors.
> 
> `{'msg': 'AccessToken is expire', 'traceId': 'e0aaf937bbef4e8bab8b14dfab451d34', 'code': 'auth.token.expire'} `
> 
> `{'msg': 'Failed to convert value of type \'java.lang.String\' to required type \'java.lang.Long\'; nested exception is java.lang.NumberFormatException: For input string: "None"', 'traceId': 'a7aa94bed56e4de697dba755c17adef4', 'code': '400', 'success': False}`

I see. What you did to get this result is correct; I was able to verify the output by implementing the code statements below:
![image](https://user-images.githubusercontent.com/121399099/210086334-5ddb12bb-2604-4de4-adfb-16087a9127bc.png)
![image](https://user-images.githubusercontent.com/121399099/210085900-a5365036-a256-43ff-9f08-78661f41d242.png)

You might need to consider restarting your cells and clearing all outputs on Jupyter before continuing. Also, you might need to refer to the accesstoken value to see if it had changed since the last time you obtained it.

**Comment by osaidz** [comment]
@tradingwithme  

> You might need to consider restarting your cells and clearing all outputs on Jupyter before continuing.

I think this was the issue. Just assigning the wb._access_token directly is sufficient to get quotes.  

Yesterday, I was able to login once using `wb.login(wb_user, wb_pass)` it prints  'refreshtoken, 'tokenExpireTime', That's how I got the values for api_login(...). 

It is no longer working and results in "Fail to verify image verification code" again.
